site/_posts/2018-02-12-discovering-a-ruby-memory-leak.html.md

```diff
@@ -0,0 +1,441 @@
+are the good parts!
+
+I ended up running across [this blog post][4] by Nelson Elhage
+([@nelhage](https://github.com/nelhage)), which ended up giving us an idea
+how to use `gdb` to hook into an existing process and run some arbitrary ruby
+code in the existing environment.
+
+The technique described in the blog post to find that memory leak was actually
+to take a look at the core dumps as hex and determine where patterns of leaked
+memory were.  We weren't that smart to be able to understand that ourselves or
+make use of it properly in our case, but we did end up using the
+`rb_eval_string("")` portions to our advantage though, and that ended up being
+super useful.
+
+At the time, we were under the impression that `MiqServer#sync_workers` was the
+culprit, so we ended up attaching to an existing 'warmed up' server process
+that had been running for a bit so most of the code it would load would already
+be autoloaded.  With that, we tried running a high number of invocations of
+that method in a row, and watched the `/proc/<PID>/smaps` to see if the
+`[heap]` section grew.  It looked something like this\*\*\*:
+
+```console
+$ gdb -p [PID_OF_THE_MIQ_SERVER_PROCESS]
+(gdb) p rb_eval_string("GC.start")
+(gdb) p rb_eval_string("500.times { MiqServer.my_server.sync_workers; GC.start }")
+```
+
+Nothing.  Not even a budge.  Well shoot... there goes that theory.  
+
+At this point, we looked back to the servers we had running that were modified
+with the  known "non-leaking" portions commented out, and checked if there was
+anything else that we might have been missing.  We noticed that in addition to
+`MiqServer#sync_workers` still being uncommented, a call to
+`MiqWorker.status_update_all` that was directly preceding it was also still
+being executed.
+
+Now, on it's own, this method looks seemed pretty tame. Effectively it:
+
+1. Uses basic ActiveRecord to fetch all the workers under the current server 
+2. Makes a call to our own code to fetch usage data for the given worker pids
+3. Saves the data to the DB records for internal reporting/monitoring
+
+The first and last bits were ruled out, and we were pretty sure we already
+tested the middle bit via some scripts outside of the context of ManageIQ, but
+for verbosity sake, using the same `gdb` process, we decided to update the
+`500.times {}` block to include that call:
+
+```ruby
+MiqServer.my_server.sync_workers; MiqWorker.status_update_all; GC.start 
+```
+
+And... it leaked... !?
+
+
+
+Step 5:  Wait... wut!? (aka:  replicated it without our app)
+------------------------------------------------------------
+
+So, at this point, we started to look at the call to
+`MiqProcess.processInfo(worker_pid)` in `MiqWorker.status_update_all` since the
+other areas are mostly ActiveRecord.  Besides, it's easier to blame our own
+libraries. ;)
+
+Now, before you say "good move dummies, that was your code... why didn't try
```

Comment:
Don't make assumptions like this from your reader. I wasn't thinking that.

Reply:
Edit can be found here:

https://gist.github.com/NickLaMuro/730b252a48019cc6bef85270ac16ff44/revisions#diff-4756561d24941a282634640442f5d8be

Commit message contents:

I mulled this over for a bit, and decided not to change it TOO much, but I did make it less abrasive since I think it retains the same effect with the wording and avoids possibly offending our soft shelled exterior developer @chrisarcand.
    
Now, I think the comment below was also getting at the fact that this might be repeating a bit of info and the intent was also to remove some clutter from the post.  I disagree, if that was the case, since I think it was a terse device in trying to explain that we had tried this before, but failed, and the next paragraph helps re-enforce this point without directly saying that.
    
And while the last sentence paragraph above does seem similar, I think it is accomplishing a different goal of establishing rational for our new plan of attack, versus this sentence, which is going over a fact that was glossed over in a previous section:
    
> These scripts can be found [here][9] and mostly failed to reliably recreate the leak.